### PR TITLE
Add union of replacer function and string to lib d.ts

### LINKED ISSUE #29789: `.call` selects the wrong overload for `String.prototype.replace`
**TypeScript Version:**  v3.3.1

**Search Terms:** call replace overload

**Code**

```ts
String.prototype.replace.call(
    'one string',
    /a/g,
    'two string', // this line errors
);
```

**Expected behavior:**
No error.

**Actual behavior:**
errors with `error TS2345: Argument of type '"two string"' is not assignable to parameter of type '(substring: string, ...args: any[]) => string'`

**Playground Link:** http://www.typescriptlang.org/play/#src=String.prototype.replace.call(%0D%0A%20%20%20%20'one%20string'%2C%0D%0A%20%20%20%20%2Fa%2Fg%2C%0D%0A%20%20%20%20'two%20string'%2C%0D%0A)%3B but the error doesn't seem to show up there.

**Related Issues:** no

### PR DESCRIPTION
**Pull request by orta**
Fixes #29789 - replacing https://github.com/microsoft/TypeScript/pull/34941 which couldn't be migrated to main ( https://github.com/microsoft/TypeScript/issues/44323 )

### CODE REVIEW

**Review by jaceppo — APPROVED** [review]

File: src/lib/es5.d.ts [hunk 1]
```diff
@@ -433,6 +433,13 @@ interface String {
      */
     replace(searchValue: string | RegExp, replacer: (substring: string, ...args: any[]) => string): string;
 
+    /**
+     * Replaces text in a string, using a regular expression or search string.
+     * @param searchValue A string to search for.
+     * @param replacer A string containing the text to replace for every successful match or a function that returns the replacement text.
+     */
+    replace(searchValue: string | RegExp, replacer: string | ((substring: string, ...args: any[]) => string)): string;
```
Comment (sandersn) [inline src/lib/es5.d.ts:441]: @orta do you know why 3 overloads of `replace` are needed? It seems like this new one is a supertype of the previous 2, even in es2015.symbol.wellknown, so maybe only the original two can be dropped.

I may be getting the variance backward in es2015.symbol.wellknown, though; could that be why there are 3 overloads?

**Review by sandersn — CHANGES_REQUESTED** [review]
I'd like to know why there are 3 overloads, and whether 1 would do.

File: src/lib/es5.d.ts (hunk 1 above)
Comment (SamB) [reply src/lib/es5.d.ts:441]: Well, if it weren't for those pesky users, one overload in es5.d.ts and one in es2015.symbol.wellknown.d.ts would be plenty. I believe it's about sparing them the cognitive load of parsing through the union *and* parsing through the doc text *and* matching up the alternatives in the prose with the alternatives in the type.

Incidentally, these docs are already quite confusing, and the first existing overload in es2015.symbol.wellknown looks to suffer from "copy/paste/forget to update". (I wrote a patch at https://github.com/microsoft/TypeScript/commit/fbb49395960fb592c1caa90187262e6003fd0110 but when I went to PR I found out I needed an issue number, so I was looking for existing issues and that's how I ended up here.)

But the part I really found confusing was "for every successful match": I guess it's technically true, but what it doesn't say is that only one match is even *attempted* for strings, or for any RegExps that don't have the `g` flag.

I suppose I should go report an issue ...

**Comment by sandersn** [comment]
I'm going to close this PR. I'm extremely wary of touching commonly used types in the lib, since any change ends up breaking somebody, and I don't think the improvement is worth it. 
I'd rather see a fix just for the documentation as in #47969.

**Pull request closed without merging**